### PR TITLE
Refactor `ferrocene-self-test`

### PR DESCRIPTION
**Pull request by Urhengulas**

### CODE REVIEW

File: ferrocene/tools/self-test/src/env.rs [hunk 1]
```diff
@@ -25,11 +27,19 @@ pub(crate) struct Env {
...
 }
 
 impl Env {
-    pub(crate) fn gather() -> Self {
-        Self { path: std::env::var_os("PATH") }
+    pub(crate) fn gather() -> Result<Self, Error> {
+        Ok(Self {
+            path: std::env::var_os("PATH").ok_or_else(|| Error::CCompilerNotFound {
+                error: crate::error::FindBinaryInPathError::NoEnvironmentVariable,
+            })?,
```
[the remark this answers is not in the document]
Comment (Urhengulas) [reply ferrocene/tools/self-test/src/env.rs:40]: Well, actually semantics do change, but I think it is better. I unset `PATH` and it behaves like following:

```console
$ # before
$ ./bin/ferrocene-self-test 
    Info: using sysroot /home/urhengulas/Documents/ferrocene
 Success: binary rustc is valid
 Success: binary rustdoc is valid
 Success: binary cargo is valid
 Success: target installed correctly: x86_64-unknown-linux-gnu
 Success: target installed correctly: armv8r-none-eabihf
 Success: bundled linker detected
 Success: bundled linker-wrapper detected
   Error: unable to find LLD-compatible C compiler for `x86_64-unknown-linux-gnu`
    Note: the error code is FST_023
$ # after
$ ./bin/ferrocene-self-test2
   Error: C compiler not found on the system
   Cause: the PATH environment variable is not set
    Note: the error code is FST_011
```

The first one does not even mention `PATH`.

**Review by Hoverbear — APPROVED** [review]
Seems a nice overall improvement.

**Comment by tshepang** [comment]
`x test ferrocene/tools/self-test` fails

File: ferrocene/tools/self-test/src/targets.rs [hunk 2]
```diff
@@ -52,6 +52,15 @@ impl Deref for Target {
     }
 }
 
+impl TargetSpec {
+    fn expected_libs(&self) -> &[&str] {
```
Comment (tshepang) [inline ferrocene/tools/self-test/src/targets.rs:56]: me would keep this inlined
- there is only one user
- is trivial

Comment (Urhengulas) [reply ferrocene/tools/self-test/src/targets.rs]: Reverted it

**Comment by Urhengulas** [comment]
> `x test ferrocene/tools/self-test` fails

Fixed the tests

File: ferrocene/tools/self-test/src/env.rs [hunk 3]
```diff
@@ -18,18 +20,55 @@ pub(crate) const SELFTEST_RUST_HASH: Option<&str> = option_env!("SELFTEST_RUST_H
...
 }
 
 impl Env {
-    pub(crate) fn gather() -> Self {
-        Self { path: std::env::var_os("PATH") }
+    pub(crate) fn gather() -> Result<Self, Error> {
+        Self::new(
+            std::env::var_os("PATH"),
+            std::env::var("FST_PRINT_DETAILED_ARGS").ok(),
+            std::env::var("FST_PRINT_DETAILED_ERRORS").ok(),
```
Comment (pietroalbini) [inline ferrocene/tools/self-test/src/env.rs:41]: We should just store a boolean of whether the variable is present or not, as that's all the only way we are using those variables.

Comment (Urhengulas) [reply ferrocene/tools/self-test/src/env.rs]: Done

**Comment by tshepang** [comment]
@Urhengulas can you remove the revert commit from history... keep it clean

**Comment by tshepang** [comment]
would also be kool to have the tests pass for each commit that would make them fail (instead of fix them in a later commit)

**Review by tshepang — APPROVED** [review]
bors merge

**Pull request merged**